### PR TITLE
Command "show-generation" Becomes "diff" + Behaviour From Spec.

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch:
- Replaces the `show-generation` command with the `diff` command.
- Performs attendant file, identifier and comment name modifications.
- Allows the operator to supply the _--all_ flag, which influences whether unit tracking detail is shown.
- Allows omission of branch name to get information from all active branches.
- Uses the branch configuration deltas to show new values for _changed_ charm configuration settings.

## QA steps

- `export JUJU_DEV_FEATURE_FLAGS=generations`
- Bootstrap.
- `juju deploy redis -n 2`.
- `juju branch test-branch`.
- `juju diff` shows:
```
test-branch:
  created: 12 Apr 2019 14:13:25+02:00
  created-by: admin
  applications: []
```
- `juju track test-branch redis/0`
- `juju diff` shows:
```
test-branch:
  created: 12 Apr 2019 14:13:25+02:00
  created-by: admin
  applications:
  - application: redis
    progress: 1/2
    config: {}
```
- `juju diff --all` shows:
```
test-branch:
  created: 12 Apr 2019 14:13:25+02:00
  created-by: admin
  applications:
  - application: redis
    progress: 1/2
    units:
      tracking:
      - redis/0
      incomplete:
      - redis/1
    config: {}
```
- `juju config redis databases=4 --branch test-branch`.
- `juju config redis password=pass --branch master`.
- `juju config redis --reset password --branch test-branch`.
- `juju diff test-branch` shows:
```
test-branch:
  created: 12 Apr 2019 14:13:25+02:00
  created-by: admin
  applications:
  - application: redis
    progress: 1/2
    config:
      databases: 4
      password: ""
```
- `juju diff too many` shows `ERROR expected at most 1 branch name, got 2 arguments`.

## Documentation changes

None

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Command "show-generation" becomes "diff". Config delta to be sourced from branch document.](https://trello.com/c/W9c5OUiX/1266-command-show-generation-becomes-diff-config-delta-to-be-sourced-from-branch-document)

File: state/application.go [hunk 1]
```diff
@@ -1372,9 +1372,7 @@ func (a *Application) UpdateApplicationSeries(series string, force bool) (err er
 				{"unitcount", a.doc.UnitCount}},
 			Update: bson.D{{"$set", bson.D{{"series", series}}}},
 		}}
-		if err != nil {
```
Comment (manadart) [inline state/application.go]: This `err` is always nil, so this block was redundant.

File: state/application.go [hunk 2]
```diff
@@ -1726,13 +1724,9 @@ func (a *Application) addUnitOpsWithCons(args applicationAddUnitOpsArgs) (string
 	// history entries. This is risky, and may lead to extra entries, but that's
 	// an intrinsic problem with mixing txn and non-txn ops -- we can't sync
 	// them cleanly.
-	if unitStatusDoc != nil {
```
Comment (manadart) [inline state/application.go]: Again here, all of these conditions are always true.

We could follow the logic in these methods better is we didn't let them get so big.

File: state/application.go [hunk 3]
```diff
@@ -1741,7 +1735,7 @@ func (a *Application) addUnitOpsWithCons(args applicationAddUnitOpsArgs) (string
 func (a *Application) addUnitSubordinateCharmProfileOp(subName, principalName string, profile *charm.LXDProfile) ([]txn.Op, error) {
 	// Because this is not part of a charm upgrade path, it is okay to
 	// short circuit here.
-	if profile == nil || (profile != nil && profile.Empty()) {
```
Comment (manadart) [inline state/application.go]: Another tautology I couldn't resist removing.

Sorry for the noise from these - it's just in the 1 file/commit.

File: cmd/juju/model/diff.go [hunk 4]
```diff
@@ -69,8 +76,8 @@ func (c *showGenerationCommand) Info() *cmd.Info {
 	info := &cmd.Info{
 		Name:    "show-generation",
```
[the remark this answers is not in the document]
Comment (manadart) [reply cmd/juju/model/diff.go]: Oops. Thanks for the catch.

**Comment by manadart** [comment]
@hmlanigan I think I must have had some un-pushed commits here. A bunch of these were already addressed; apologies.

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 5]
```diff
@@ -31,7 +31,7 @@ type API struct {
 // NewModelGenerationFacade provides the signature required for facade registration.
 func NewModelGenerationFacade(ctx facade.Context) (*API, error) {
```
[the remark this answers is not in the document]
Comment (manadart) [reply apiserver/facades/client/modelgeneration/modelgeneration.go]: Not sure. If the consensus from above ends up being to abandon "generation" nomenclature, then yes.

**Comment by mitechie** [comment]
Tried to QA but ran into this:

    juju branch testing
    Active branch set to "testing"
    juju diff
    ERROR unrecognized command: juju diff

I did set the FF before bootstrap (and branch worked so I assumed that would help verify)

    export JUJU_DEV_FEATURE_FLAGS=generations

**Comment by manadart** [comment]
Yes, I created a card for tab-completion today :)

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**